Resources system initial design

Teifion opened the pull request.

Avaristimo commented on design docs/Resources system initial design.md:
```diff
@@ -57,10 +58,8 @@ Composite resources represent a resource with two or more components making up t
 These compositions can vary in both content (one asteroid might have carbon and iron, another might have oxygen and iron) and in quantities (one is rich in iron, another poor). We will need a way to represent this.
 
 ### Composite types
```
Do we want to differentiate between resources and other ingame "stuff"?
ferrous ore -> iron -> steel sheets -> ship components -> spaceship

should these be separated into resources and then transformed into other entities, or all be the same with different tags?

Teifion replied:
Resolved in a VC discussion. We will use large amounts of ore to make things thus allowing them to be grouped in this system.

Teifion commented:
@Avaristimo I have moved Quality out of the scope of the system at this stage and updated the document based on our conversation earlier. Please feel free to merge if you approve of it but equally feel free to request changes.

Teifion commented on design docs/Resources system initial design.md:
```diff
@@ -49,8 +51,52 @@ Representing a specific instance of a resource existing. If two instances are co
-	- Composites need to have the ability to have their contents hidden, e.g. you have muddy water but you don't know what the mud is
+# Things still unceratin
+- Composites need to have the ability to have their contents hidden, e.g. you have muddy water but you don't know what the mud is. This strikes me as something we can do after implementing the resource system as opposed to being part of it. Gut feeling tells me it's the same pattern as an authorisation issue (permissions come from doing scans etc).
+- Do we need both mass and volume?
```
Keeping them in.

Teifion commented:
> quality would add a ton of complexity at this point that's not yet needed and can probably be added relatively easily later

Yeah, that's why I moved it to a "we want to do this in the future" rather than making it a part of the thing now.

dormouse-md commented on design docs/Resources system initial design.md:
```diff
@@ -0,0 +1,144 @@
+A [[Resources system]] underpins almost every aspect of an economic system and several more to boot. We want to have a flexible resource system allowing for complexity and depth without requiring either. A list of functional behaviour we want to enable is:
+- Presence in a physical location (e.g. station, ship etc)
+- Transfer resources from one storage to another at the same location (e.g. ship docked at a station)
+- Transport resources between locations by having the container (e.g. ship) travel
```
Any smaller transporting instances would be considered - like backpack, trolley, crate, palette etc or  mostly  thinking of bigger picture here?

Teifion replied:
If it adds value I'd be happy for a greater level of detail. What sort of effects are you imagining?

dormouse-md replied:
I was thinking about transition in between mining and ships, trading, smuggling, keeping souvernirs etc. But prorbably it's too soon. So happy with the level of detail present

dormouse-md commented on design docs/Resources system initial design.md:
```diff
@@ -0,0 +1,144 @@
+- Transfer resources from one storage to another at the same location (e.g. ship docked at a station)
+- Transport resources between locations by having the container (e.g. ship) travel
+- Varying storage requirements such as mass, volume and specialised containers (though not limited to just these)
+- Affecting of resources over time (e.g. [[spoilage]])
```
> Would contamination/loss also fall under this category? ie hole in a water tank, leak in a oxigen one, etc

Never mind, got to the Spoilage section

dormouse-md commented on design docs/Resources system initial design.md:
```diff
@@ -0,0 +1,144 @@
+| Plants     | 400  | 100    | needs-atmosphere, delicate |
+
+### Resource instances
+Representing a specific instance of a resource existing. If two instances are combined then one would be deleted. It is expected within the database the instances will be further broken down by location type (e.g. ship or station).
```
What type of the database are you planning on using?(Sorry it's probably somewhere in the docs which  I haven't gone through yet)

Teifion replied:
Currently using postgres, plan to use that until it breaks at which stage looking at other options. Caching within Elixir is incredibly easy so I plan to do that a bit as we go.

dormouse-md replied:
I used postgres for many different projects and usually it was enough. Let see how it performs here!

dormouse-md commented on design docs/Resources system initial design.md:
```diff
@@ -0,0 +1,144 @@
+| Iron sheet | 1000     |
+
+## Composite resources
+Composite resources represent a resource with two or more components making up the actual content of the resource. For example, extracting ferrous ores will yield you iron but there will be other elements there too. These will be present and depending on equipment and resources can be identified (or possibly be known from the outset if surveys take place).
```
When we were talking with @Avaristimo he mentioned that you guys want to enchourage collaboration as much as possible, what if joined mining/extraction or making of something would yeald more (like multiplied on coefficient)/take less time/mining or making special resources would  be possible only when making something together?
Sorry if it is a wrong place to ask :)

Teifion replied:
Great idea and I have a few ideas on ways to do that; I want to keep the scope of this design to just resources. I would expect a mining/extraction system to be in the near future and that's exactly where this would go.

I think for the level of granularity of the resource system the co-op benefits will come from the way it's used rather than be codified into the design. E.g. you hire an ally to transport goods because they have a bigger/faster ship.

dormouse-md replied:
Perfect, thank you!

dormouse-md commented on design docs/Resources system initial design.md:
```diff
@@ -0,0 +1,144 @@
+| Ferrous Ore | 800, 100, 50 | 1000     | 1300          | 100             |
+
+## Tagged behaviours
+The purpose of the tags is to allow overlapping behaviours to be defined for different resource types. It is intended new behaviours can be added without overhauling the entire system without needing to plan out it completely at this stage.
```
Like a group, right?

Teifion replied:
Yes.

dormouse-md commented on design docs/Resources system initial design.md:
```diff
@@ -0,0 +1,144 @@
+
+### Spoilage
+Spoilage is the function of transforming a resource from one type to another given time and conditions. Some examples include:
+- Food spoiling if not refrigerated or frozen
```
Where would food come from? Grown on ship?
Also what about oxygen?

Teifion replied:
Outside the scope of the resources system (instead it'd be the extraction/production system) but I'd imagine food could be created in a number of different ways with different tradeoffs. I would also expect there to be different types of food to add complexity!

dormouse-md replied:
Makes sense! Thank you!

dormouse-md commented on examples/shows and movies/The Expanse.md:
```diff
@@ -1,10 +1,13 @@
-aliases:
+aliases: 
+websites:
+  - https://en.wikipedia.org/wiki/The_Expanse_(novel_series)
```
It's amazing! 
Maybe The Martian can be  of any help?

Teifion replied:
As soon as it's referenced by something it'll be added. Feel free to create a PR adding examples of things that are fleshed out.

dormouse-md reviewed (commented):
Looks really cool!
Added a few questions, sorry if it is too much
